### ISSUE TITLE
Static range with disconnected (non-same-tree) endpoints is considered valid

### ISSUE THREAD

**Issue by bzbarsky** [w3c/staticrange#14]
With the changes made in https://github.com/w3c/staticrange/issues/8 clarifying what https://w3c.github.io/staticrange/#staticrange-position-is-less-than does, I can now evaluate what https://w3c.github.io/staticrange/#staticrange-is-valid is doing.  Consider a StaticRange whose endpoints are not in the same DOM tree.  "sr’s end position is less than sr’s start position" will test false, because https://dom.spec.whatwg.org/#concept-tree-preceding returns false in this case.  So the "is valid" algorithm will return true.  This doesn't seem right.

**Comment by garykac** [comment]
Yes, that's certainly not right. The assumption was that the nodes were in the same DOM tree, but that should be made explicit.

What do you think about the following fix:

(1) Update "is valid" to first ensure that the 2 nodes are in the same DOM tree.
(2) Update the text to `throw` whenever an invalid `StaticRange` would be created.

With (1), we'll never call "is less than" or "is equal to" with Nodes in different DOM trees.  We could still have a DOM-tree check, but it seems unnecessary.

For (2), I'll need to add text describing how the constructor works (and when it throws).

Given that we've agreed that we should `throw` on invalid ranges, I think it would be cleaner if we make the xContainer and xOffset attributes `readonly` and have `setStart` and `setEnd` methods which would `throw`.  Otherwise, we'd need to `throw` in response to updating the attributes, which doesn't feel right to me.

Any thoughts on this approach?

**Comment by bzbarsky** [comment]
What happens if a StaticRange is created with endpoints that are in the same DOM tree, then one of them is removed from that tree?  It seems to me that you can prevent creation of StaticRange that's invalid at creation time, but you can't prevent an already-valid one from becoming invalid as the DOM mutates.

So this is more about catching creation errors than enforcing a validity invariant, right?

**Comment by rniwa** [comment]
Yeah, we most certainly need to update the definition of *is valid* to check that start & end containers belong to the same tree (probably by checking that their root node is identical).

**Comment by garykac** [comment]
I updated the text in https://github.com/w3c/staticrange/commit/f2af0f4d240eccc1df1030c7b982763d545df8f4 to check that the 2 containers share a common ancestor.

**Comment by annevk** [comment]
You need to use inclusive ancestor if you want to go that route, but checking root seems better.

**Comment by johanneswilm** [comment]
@annevk So what should the exact sentence be? Something like:
 
       If sr's startContainer's root and sr's endContainer's root are not the same

or

       If sr's startContainer and endContainer do not share a common inclusive ancestor

Would either of these work or should it be something entirely different?

**Comment by annevk** [comment]
The former seems good, though I'd use "is" to compare them. I also think though that if you were to patch this against DOM you'd factor this out as a shared primitive.

**Comment by johanneswilm** [comment]
Great!

> I also think though that if you were to patch this against DOM you'd factor this out as a shared primitive.

Basically this is about whether to introduce a new term for describe two elements having a common root, right? That sounds like a question related mostly to how one wants to organize the DOM specs. The DOM maintainers may have stronger opinions about than those of us only concerned with static ranges, so maybe this could be left for the DOM maintainers to figure out?

**Comment by annevk** [comment]
As editor of the DOM Standard that's the feedback I'd give on a PR.

**Comment by johanneswilm** [comment]
@annevk, Ok. So, in case Gary decides to go with "common inclusive ancestor": In the both versions of the DOM spec, one can find the phrase "common inclusive ancestor" (twice in a note).  Would it make sense to rephrase those lines so that "common inclusive ancestor" becomes a defined term and move those two lines into the tree part of the terminology section (1.1/2.1)?

Alternatively, if he wants to go with the other one, would you prefer a new term to be defined to describe a situation where two nodes don't have a common root or is there already a term for that situation that then only needs to be defined? If you want to use "is" instead of "are", I guess the text will have to be something like:

        If sr's startContainer's root is not the same as sr's endContainer's root

**Comment by annevk** [comment]
I haven't looked at this in enough detail to really say, but I suspect that maybe making https://dom.spec.whatwg.org/#dom-range-commonancestorcontainer use a common algorithm is the way to go (tracked by https://github.com/whatwg/dom/issues/56 at the moment). That new algorithm could return null if there is no common ancestor, making it suitable for reuse here.

**Comment by johanneswilm** [comment]
Ok, looking at that issue, that seems to be a longer process that will take years to resolve. Agreeing on an algorithm to determine the lowest common ancestor is maybe not that simple, but checking whether two root nodes are the same should be straightforward.

**Comment by annevk** [comment]
I don't think it should take years really. It's a "good first issue" which I try to avoid touching so new contributors get a chance to fix things up.

**Comment by annevk** [comment]
Can someone tell me what callers need this algorithm?

I looked at https://w3c.github.io/input-events/index.html and it doesn't really define how any of the `StaticRange` stuff is supposed to work.
